Update getting started documentation (#79)
terraform-google-modules/terraform-google-project-factory: docs/TROUBLESHOOTING.md

## Terminology

See [GLOSSARY.md][glossary].

- - -
## Problems

* [Common issues](#common-issues)
* [Seed project missing APIs](#seed-project-missing-apis) - The seed project is
  missing required APIs.
* [Service account missing roles](#service-account-missing-roles) - The service
  account has insufficient permissions.

- - -
### Common issues

* [Unable to query status of default GCE service
  account](#unable-to-query-status-of-default-gce-service-account)
* [Cannot manage G Suite group membership](#cannot-manage-g-suite-group-membership)

#### Unable to query status of default GCE service account

> When the Project Factory is used with a misconfigured seed project, it partially

Author: Amended

> generate a new project, fail, and get into a state where it can no longer

Author: Amended

> generate plans.

Author: Amended

## Terminology

See [GLOSSARY.md][glossary].

- - -
## Problems

* [Common issues](#common-issues)

> * [Seed project missing APIs](#seed-project-missing-apis) - The seed project is

Reviewer: ```suggestion
* [Seed Project missing APIs](#seed-project-missing-apis) - The Seed Project is
```

  missing required APIs.

> * [Service account missing roles](#service-account-missing-roles) - The service

Reviewer: ```suggestion
* [Seed Service Account missing roles](#seed-service-account-missing-roles) - The Seed Service
```

>   account has insufficient permissions.

Reviewer: ```suggestion
  Account has insufficient permissions.
```

- - -
### Common issues

* [Unable to query status of default GCE service
  account](#unable-to-query-status-of-default-gce-service-account)
* [Cannot manage G Suite group
  membership](#cannot-manage-g-suite-group-membership)

#### Unable to query status of default GCE service account

When the Project Factory is used with a misconfigured Seed Project, it will
partially generate a new Target Project, fail, and enter a state where it can no
longer generate Target Projects.

**Error message:**

```
Error: Error refreshing state: 1 error(s) occurred:

* module.project-factory.data.google_compute_default_service_account.default: 1
  error(s) occurred:

* module.project-factory.data.google_compute_default_service_account.default:
  data.google_compute_default_service_account.default: Error reading GCE service
  account not found: googleapi: Error 403: Project 449949713944 is not found and
  cannot be used for API calls. If it is recently created, enable Compute Engine
  API by visiting
  https://console.developers.google.com/apis/api/compute.googleapis.com/overview?project=449949713944
  then retry. If you enabled this API recently, wait a few minutes for the

>   action to propagate to our systems and retry., accessNotConfigured ```

Reviewer: ```suggestion
  action to propagate to our systems and retry., accessNotConfigured
```

Reviewer: The ``` needs to be moved to the next line.

        ```
    1. Associate a billing account with the seed project:

>         ```

Reviewer: ```suggestion
        ```sh
```

        # Requires `roles/billing.projectManager` on $SEED_PROJECT and
        # `roles/billing.user` on the billing account
        gcloud alpha billing projects link "$SEED_PROJECT" --billing-account=[billing-account]
        ```

> 1. Enable `compute.googleapis.com` on the target project:

Reviewer: ```suggestion
1.  Enable `compute.googleapis.com` on the Target Project:
```

>     ```

Reviewer: ```suggestion
    ```sh
```

>     # Requires `roles/serviceusage.admin` on $TARGET_PROJECT

Reviewer: ```suggestion
    # Requires `roles/serviceusage.admin` on $TARGET_PROJECT
```

>     gcloud services enable compute.googleapis.com --project $TARGET_PROJECT

Reviewer: ```suggestion
    gcloud services enable compute.googleapis.com --project $TARGET_PROJECT
```

    ```
    This should be run in the context of the Seed Service Account.  You can add

>     the service account to your list of authentication credentials by issuing

Reviewer: ```suggestion
    the Seed Service Account to your list of authentication credentials by issuing
```

>     the following command and importing your service account key:

Reviewer: ```suggestion
    the following command and importing the Seed Service Account key:
```

>     ```

Reviewer: ```suggestion
    ```sh
```

**Cause:**

By default service accounts cannot manage G Suite resources, because the G Suite
directory API requires a user account with admin rights. A service account can
impersonate a user account with admin rights to interact with the directory API,
but the service account has to be granted domain wide delegation rights to
impersonate users, and must be granted OAuth scopes for the functionality it
needs to interact with.

> If domain wide delegation on the appropriate is not granted to the service

Reviewer: ```suggestion
If domain wide delegation is not granted to the Seed Service
```

> account, it will fail to get the access token needed to interact with the

Reviewer: ```suggestion
Account, it will fail to obtain the access token needed to interact with the
```

Directory API and fail.

See the [README G Suite documentation](../README.md#g-suite) for more
information.

**Solution:**

Enable [domain wide
delegation](https://developers.google.com/admin-sdk/directory/v1/guides/delegation)

> on the service account with the following scopes:

Reviewer: ```suggestion
on the Seed Service Account with the following scopes:
```

* https://www.googleapis.com/auth/admin.directory.group
* https://www.googleapis.com/auth/admin.directory.group.member

- - -
### Seed project missing APIs

> The Project Factory requires the following services are enabled on the seed

Reviewer: ```suggestion
The Project Factory requires the following services to be enabled on the Seed
```

> project. If these APIs are not enabled the Project may fail to generate

Reviewer: ```suggestion
Project. If these APIs are not enabled, the Project Factory may fail to generate
```

resources, or generate incomplete resources.

A canonical list of required APIs is available in the
[README](../README.md#apis).

* [cloudresourcemanager.googleapis.com](#missing-api-cloudresourcemanagergoogleapiscom)
* [cloudbilling.googleapis.com](#missing-api-cloudbillinggoogleapiscom)
* [iam.googleapis.com](#missing-api-iamgoogleapiscom)
* [appengine.googleapis.com](#missing-appenginegoogleapiscom) (when managing an

>   App Engine instance on the target project)

Reviewer: ```suggestion
  App Engine instance on the Target Project)
```

* module.project-factory.data.google_organization.org: 1 error(s) occurred:

* module.project-factory.data.google_organization.org:
  data.google_organization.org: Error reading Organization Not Found :
  [organization-id]: googleapi: Error 403: Cloud Resource Manager API has not
  been used in project [seed-project-number] before or it is disabled. Enable it
  by visiting
  https://console.developers.google.com/apis/api/cloudresourcemanager.googleapis.com/overview?project=[seed-project-number]
  then retry. If you enabled this API recently, wait a few minutes for the
  action to propagate to our systems and retry., accessNotConfigured

```

**Cause:**

> The seed project does not have the `cloudresourcemanager.googleapis.com` API

Reviewer: ```suggestion
The Seed Project does not have the `cloudresourcemanager.googleapis.com` API
```

    ```

> * **Option 2:** create a new service account and enable the required APIs:

Reviewer: ```suggestion
* **Option 2:** create a new Seed Service Account and enable the required APIs:
```

>     ```

Reviewer: ```suggestion
  ```sh
```

>     # requires `roles/resourcemanager.organizationAdmin` on the organization and

Reviewer: ```suggestion
  # requires `roles/resourcemanager.organizationAdmin` on the organization and
```

>     # `roles/serviceusage.serviceUsageAdmin` on $SEED_PROJECT

Reviewer: ```suggestion
  # `roles/serviceusage.serviceUsageAdmin` on $SEED_PROJECT
```

>     ./helpers/setup-sa.sh [organization id] "$SEED_PROJECT"

Reviewer: ```suggestion
  ./helpers/setup-sa.sh [organization id] "$SEED_PROJECT"
```

* module.project-factory.google_project.project: 1 error(s) occurred:

* google_project.project: Error setting billing account "[billing-account]" for
  project "projects/[target-project-id]": googleapi: Error 403: Cloud Billing
  API has not been used in project [seed-project-number] before or it is
  disabled. Enable it by visiting
  https://console.developers.google.com/apis/api/cloudbilling.googleapis.com/overview?project=[seed-project-number]
  then retry. If you enabled this API recently, wait a few minutes for the
  action to propagate to our systems and retry., accessNotConfigured
```

**Cause:**

> The seed project does not have the `cloudbilling.googleapis.com` API enabled.

Reviewer: ```suggestion
The Seed Project does not have the `cloudbilling.googleapis.com` API enabled.
```

**Notes:**

> This error will occur *once* when applying a Terraform plan when the **seed

Reviewer: ```suggestion
This error will occur *once* when applying a Terraform plan when the Seed
```

> project** does not have the Cloud Billing API enabled. On subsequent Terraform

Reviewer: ```suggestion
Project does not have the Cloud Billing API enabled. On subsequent Terraform
```

runs, Terraform will generate an error indicating that the Compute Engine API is

> not enabled on the **target project**. Watch out for this!

Reviewer: ```suggestion
not enabled on the Target Project. Watch out for this!
```

* google_service_account.extra_service_account: 1 error(s) occurred:

* google_service_account.extra_service_account: Error creating service account:
  googleapi: Error 403: Identity and Access Management (IAM) API has not been
  used in project [seed-project-number] before or it is disabled. Enable it by
  visiting
  https://console.developers.google.com/apis/api/iam.googleapis.com/overview?project=[seed-project-number]
  then retry. If you enabled this API recently, wait a few minutes for the
  action to propagate to our systems and retry., accessNotConfigured

* module.project-factory.google_service_account.default_service_account: 1
  error(s) occurred:

* google_service_account.default_service_account: Error creating service
  account: googleapi: Error 403: Identity and Access Management (IAM) API has
  not been used in project [seed-project-number] before or it is disabled.
  Enable it by visiting
  https://console.developers.google.com/apis/api/iam.googleapis.com/overview?project=[seed-project-number]
  then retry. If you enabled this API recently, wait a few minutes for the
  action to propagate to our systems and retry., accessNotConfigured
```

**Cause:**

> The seed project does not have the `iam.googleapis.com` API enabled. This

Reviewer: ```suggestion
The Seed Project does not have the `iam.googleapis.com` API enabled. This
```

    ```

> * **Option 2:** create a new service account and enable the required APIs:

Reviewer: ```suggestion
* **Option 2:** create a new Seed Service Account and enable the required APIs:
```

>     ```

Reviewer: ```suggestion
  ```sh
```

>     # requires `roles/resourcemanager.organizationAdmin` on the organization and

Reviewer: ```suggestion
  # requires `roles/resourcemanager.organizationAdmin` on the organization and
```

>     # `roles/serviceusage.serviceUsageAdmin` on $SEED_PROJECT

Reviewer: ```suggestion
  # `roles/serviceusage.serviceUsageAdmin` on $SEED_PROJECT
```

>     ./helpers/setup-sa.sh [organization id] "$SEED_PROJECT"

Reviewer: ```suggestion
  ./helpers/setup-sa.sh [organization id] "$SEED_PROJECT"
```

**Cause**:

> The App Engine API is not enabled on the seed project, which prevents creation

Reviewer: ```suggestion
The App Engine API is not enabled on the Seed Project, which prevents creation
```

> of an App Engine instance on the target project.

Reviewer: ```suggestion
of an App Engine instance on the Target Project.
```

    ```

> * **Option 2:** create a new service account and enable the required APIs:

Reviewer: ```suggestion
* **Option 2:** create a new Seed Service Account and enable the required APIs:
```

>     ```

Reviewer: ```suggestion
  ```sh
```

>     # requires `roles/resourcemanager.organizationAdmin` on the organization and

Reviewer: ```suggestion
  # requires `roles/resourcemanager.organizationAdmin` on the organization and
```

>     # `roles/serviceusage.serviceUsageAdmin` on $SEED_PROJECT

Reviewer: ```suggestion
  # `roles/serviceusage.serviceUsageAdmin` on $SEED_PROJECT
```

>     ./helpers/setup-sa.sh [organization id] "$SEED_PROJECT"

Reviewer: ```suggestion
  ./helpers/setup-sa.sh [organization id] "$SEED_PROJECT"
```

1 error(s) occurred:

* module.project-factory.gsuite_group_member.service_account_sa_group_member: 1
  error(s) occurred:

* gsuite_group_member.service_account_sa_group_member: Error creating group
  member: googleapi: Error 403: Access Not Configured. Admin Directory API has
  not been used in project 454152376537 before or it is disabled. Enable it by
  visiting
  https://console.developers.google.com/apis/api/admin.googleapis.com/overview?project=454152376537
  then retry. If you enabled this API recently, wait a few minutes for the
  action to propagate to our systems and retry., accessNotConfigured
```

**Cause:**

> The Directory Admin API is not enabled on the seed project, which prevents

Reviewer: ```suggestion
The Directory Admin API is not enabled on the Seed Project, which prevents
```

    ```

> * **Option 2:** create a new service account and enable the required APIs:

Reviewer: ```suggestion
* **Option 2:** create a new Seed Service Account and enable the required APIs:
```

>     ```

Reviewer: ```suggestion
  ```sh
```

>     # requires `roles/resourcemanager.organizationAdmin` on the organization and

Reviewer: ```suggestion
  # requires `roles/resourcemanager.organizationAdmin` on the organization and
```

>     # `roles/serviceusage.serviceUsageAdmin` on $SEED_PROJECT

Reviewer: ```suggestion
  # `roles/serviceusage.serviceUsageAdmin` on $SEED_PROJECT
```

>     ./helpers/setup-sa.sh [organization id] "$SEED_PROJECT"

Reviewer: ```suggestion
  ./helpers/setup-sa.sh [organization id] "$SEED_PROJECT"
```

    ```

- - -

> ### Service account missing roles

Reviewer: ```suggestion
### Seed Service Account missing roles
```

> The service account must have the following roles in order to fully create a

Reviewer: ```suggestion
The Seed Service Account must have the following roles in order to fully create a
```

> project factory.

Reviewer: ```suggestion
Project Factory.
```

```
* module.project-factory.data.google_organization.org: 1 error(s) occurred:

* module.project-factory.data.google_organization.org:
  data.google_organization.org: Error reading Organization Not Found :
  [organization-id]: googleapi: Error 403: The caller does not have permission,
  forbidden
```

**Cause:**

> The active service account does not have

Reviewer: ```suggestion
The Seed Service Account does not have
```

`roles/resourcemanager.organizationViewer` on the active organization. The

> organizationViewer role grants the service account the ability to look up the

Reviewer: ```suggestion
organizationViewer role grants the Seed Service Account the ability to look up the
```

GCP organization by the organization ID and fetch the associated domain name.

**Solution:**

> Use `helpers/setup-sa.sh` to create a new service account with the necessary

Reviewer: ```suggestion
Use `helpers/setup-sa.sh` to create a new Seed Service Account with the necessary
```

* module.project-factory.google_project.project: 1 error(s) occurred:

* google_project.project: error creating project [target-project-id]
  ([target-project-name]): googleapi: Error 403: User is not authorized.,
  forbidden. If you received a 403 error, make sure you have the
  `roles/resourcemanager.projectCreator` permission
```

**Cause:**

> The active service account does not have `roles/resourcemanager.projectCreator`

Reviewer: ```suggestion
The Seed Service Account does not have `roles/resourcemanager.projectCreator`
```

> on the active organization. The projectCreator role allows the service account

Reviewer: ```suggestion
on the active organization. The projectCreator role allows the Seed Service Account
```

to generate new projects within the GCP organization.

**Solution:**

> Use `helpers/setup-sa.sh` to create a new service account with the necessary

Reviewer: ```suggestion
Use `helpers/setup-sa.sh` to create a new Seed Service Account with the necessary
```

    * `roles/browser`
    * `roles/resourcemanager.projectIamAdmin`

> These roles are required for associating the target project with the host VPC

Reviewer: ```suggestion
These roles are required for associating the Target Project with the host VPC
```

and managing access to the host VPC network resources.

**Solution:**

> Use `helpers/setup-sa.sh` to create a new service account with the necessary

Reviewer: ```suggestion
Use `helpers/setup-sa.sh` to create a new Seed Service Account with the necessary
```

* module.project-factory.google_project.project: 1 error(s) occurred:

* google_project.project: Error setting billing account "[billing-account]" for
  project "projects/[target-project-id]": googleapi: Error 403: The caller does
  not have permission, forbidden
```

**Cause**:

> The service account does not have the `roles/billing.user` role on the billing

Reviewer: ```suggestion
The Seed Service Account does not have the `roles/billing.user` role on the billing
```

**Notes:**

* Granting `roles/billing.user` on the organization is not sufficient if the
  billing account is defined outside of the GCP organization. Make sure that the

>   service account has the `roles/billing.user` role on the billing account.

Reviewer: ```suggestion
Seed Service Account has the `roles/billing.user` role on the billing account.
```
